zkllvm/use-cases/zk-bridge/hashes.mdx

```diff
@@ -0,0 +1,299 @@
+
+<Tabs groupId='language'>
+  <TabItem value='cpp' label='C++'>
+    ```cpp
+    struct block_data_type {
+        typename sha2<256>::block_type prev_block_hash;
+        typename sha2<256>::block_type data;
+    };
+    ```
+  </TabItem>
+  <TabItem value='rust' label='Rust'>
+    ```rust
+    type BlockType = [Fq; 2];
+
+    struct BlockDataType {
+        prev_block_hash: BlockType,
+        data: BlockType,
+    }
+    ```
+  </TabItem>
+</Tabs>
+
+### Additional functions
+
+<Tabs groupId='language'>
+  <TabItem value='cpp' label='C++'>
+    ```cpp
+    bool is_same(
+        typename sha2<256>::block_type block0,
+        typename sha2<256>::block_type block1) {
+
+        return block0[0] == block1[0] && block0[1] == block1[1];
+    }
+    ```
+  </TabItem>
+  <TabItem value='rust' label='Rust'>
+    ```rust
+    fn is_same(x: BlockType, y: BlockType) -> bool {
+        x[0] == y[0] && x[1] == y[1]
+    }
+
+    fn hash(block1: BlockType, block2: BlockType) -> BlockType {
+        let sha = assigner_sha2_256([block1[0].0, block1[1].0], [block2[0].0, block2[1].0]);
+        [sha[0].into(), sha[1].into()]
+    }
+    ```
+  </TabItem>
+</Tabs>
+
+### Circuit function
+
+<Tabs groupId='language'>
+  <TabItem value='cpp' label='C++'>
+    ```cpp
+    [[circuit]] bool verify_protocol_state_proof (
+        typename sha2<256>::block_type last_confirmed_block_hash,
+        std::array<block_data_type, 2> unconfirmed_blocks) {
+        for (int i = 0; i < 2; i++) {
+            if (i == 0) {
+                if (!is_same(
+                    unconfirmed_blocks[i].prev_block_hash,
+                    last_confirmed_block_hash)) {
+                    return false;
+                }
+            } else {
+                typename sha2<256>::block_type evaluated_block_hash =
+                    hash<sha2<256>>(
+                        unconfirmed_blocks[i-1].prev_block_hash,
+                        unconfirmed_blocks[i-1].data);
+                if (!is_same(
+                    unconfirmed_blocks[i].prev_block_hash,
+                    evaluated_block_hash)) {
+                    return false;
+                }
+            }
+        }
+        return true;
+    }
+    ```
+  </TabItem>
+  <TabItem value='rust' label='Rust'>
+    ```rust
+    #[circuit]
+    #[unroll_for_loops]
+    fn verify_protocol_state_proof(
+        last_confirmed_block_hash: Fq,
+        unconfirmed_blocks: [BlockDataType; 2],
+    ) -> bool {
+        for i in 0..2 {
+            if i == 0 {
+                if is_same(unconfirmed_blocks[i][0], last_confirmed_block_hash) {
+                    false
+                }
+            } else {
+                let evaluated_block_hash =
+                    hash(unconfirmed_blocks[i - 1][0], unconfirmed_blocks[i - 1][1]);
+                if is_same(unconfirmed_blocks[i][0], evaluated_block_hash) {
+                    false
+                }
+            }
+            true
+        }
+    }
+    ```
```

Comment:
The code of the function is wrong. Here is the fixed version:

```rust
fn verify_protocol_state_proof(
    last_confirmed_block_hash: Fq,
    unconfirmed_blocks: [BlockDataType; 2],
) -> bool {
    for i in 0..2 {
        if i == 0 {
            if !is_same(unconfirmed_blocks[i][0], last_confirmed_block_hash) {
                return false;
            }
        } else {
            let evaluated_block_hash =
                hash(unconfirmed_blocks[i - 1][0], unconfirmed_blocks[i - 1][1]);
            if !is_same(unconfirmed_blocks[i][0], evaluated_block_hash) {
                return false;
            }
        }
    }
    true
}
```

Reply:
you're right, a mix-up on my part

Reply:
in any case, the function is now rewritten from scratch